shared/worker.go

```diff
@@ -0,0 +1,102 @@
+package darvaza
+
+import (
+	"fmt"
+	"log"
+	"sync"
+)
+
+// Worker is a routing that runs supervised
+type Worker interface {
+	Run() error
+	Cancel() error
+}
+
+// WorkGroup governs a slice of Workers
+type WorkGroup struct {
+	workers map[Worker]struct{}
+	wg      sync.WaitGroup
+	Done    chan error
+}
+
+func (s *WorkGroup) Run() error {
+	var err error
+	for k := range s.workers {
+		s.wg.Add(1)
+		go func(k Worker) {
+			defer s.wg.Done()
+			err = k.Run()
+			if err != nil {
+				select {
+				case s.Done <- err:
+				default:
+					//non blocking send
+				}
+				log.Println(err)
+				err = k.Cancel()
+				if err != nil {
+					log.Println(err)
```

Comment:
I think that this should be treated differently, at the moment I just wanted to have a rapid output and somehow this remained in code :-/. Maybe WorkGroup should have a logger member?